### ISSUE TITLE
vonmises and vonmises_line in stats: vonmises wrong and superfluous?

### ISSUE THREAD

**Issue by chrisb83** [scipy/scipy#8878]
Two distributions are currently implemented in `scipy.stats`: `vonmises` and `vonmises_line`.

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.vonmises.html
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.vonmises_line.html

According to Wikipedia (https://en.wikipedia.org/wiki/Von_Mises_distribution), the von Mises distribution is defined on [-pi, pi] (in the standard form).

`vonmises` is defined on the real line
`vonmises_line` is defined on [-pi, pi]

I don't think the implemention of `vonmises` makes sense. For instance, the `pdf` returns `np.exp(kappa * np.cos(x)) / (2*np.pi*sc.i0(kappa))` which does not converge to zero as `x` goes to infinity. The cdf also does not return sensible values, see below.

Should there only be one vonmises distribution?

### Reproducing code example:
```
from scipy import stats
vm = stats.vonmises(1)
vm.cdf(10) # 2.0281170964731907
```

### Scipy/Numpy/Python version information:
```
1.0.0 1.14.1 sys.version_info(major=3, minor=5, micro=4, releaselevel='final', serial=0)
```

**Comment by josef-pkt** [comment]
similar issue #4598

a long time (4 to 6 years) ago, Robert Cimrman and I estimated mixtures of vonmises distributions.
Anne Archibald had added the extended cdf.

**Comment by pvanmulbregt** [comment]
Which one is the "improper" one?  `vonmises`? If so then there is somewhat of a disconnect.
The "standard" Von Mises is implemented as `vonmises_line`.
The "improper" Von Mises is implemented as `vonmises`, but that is not actually a probability distribution.
 * `vonmises.cdf(-np.inf) = 0`, but `vonmises.cdf(x) ~ x/2pi` as `x -> -infinity`
* `vonmises.cdf(np.inf) = 1`, but `vonmises.cdf(x) ~ x/2pi` as `x -> infinity`
* The `cdf()` computations on the improper can make sense whenever taking the difference of two `cdf()`s, and the two `x`-values are within `2pi` of each other.  
But none of this is in the doc.

**Comment by ev-br** [comment]
I suspect this is going to keep puzzling people year after year until there is a proper circular distribution framework alongside `rv_continuous`.

**Comment by chrisb83** [comment]
I agree with pvanmulbregt that it ist confusing to call the distribution on [-pi, pi] `vonmises_line`.

I also don't understand the issue of defining the distribution on [-pi,pi] only. If one wants to evaluate the pdf / cdf at `x` where `x` is not necessarily in that interval, one would just need to map it before the pdf / cdf is applied, no?
And using `scale` and `loc`, one can define the distribution on any bounded interval [a, b]. i don't know circular distributions well, so maybe i misunderstand something....

**Comment by josef-pkt** [comment]
What might work is to make vonmises or vonmises_line into a 4 parameter distribution with 2 shape parameters, circular mean and circular dispersion, on a fixed interval [-pi, pi] that internally evaluated pdf for the correct shape given by vonmises on a shifted segment.

Users could keep loc and scale fixed, or for example use it to translate the circle to a 24 hour clock, i.e. interval [0, 24).

That might be pretty much the steps that a user needs to do currently to use vonmises as a circular distribution.

**Comment by pvanmulbregt** [comment]
Do you have an online reference for circular distributions that you recommend?

**Comment by josef-pkt** [comment]
No, or I don't remember,
My main reference was Mardia's book on Directional Statistics, otherwise I was reading articles on specific distribution families or methods like wrapping to construct those families, and most likely skimmed or selectively read some other books.

Maybe https://en.wikipedia.org/wiki/Directional_statistics#von_Mises_circular_distribution has some useful links

**Comment by pvanmulbregt** [comment]
Thanks. That makes it seem as though the usual concept of CDF from 1-d real (or real interval) distributions doesn't make sense in the circular world. There is no ordering on the domain, hence no natural endpoints, no `a`, `b` as used in `rv_continuous`.  But the measure of a circular interval `[Theta0, Theta1]` makes sense, as just the integral of the PDF between the two endpoint angles.  My interpretation of your earlier comments is that you may have used the difference `vonmises.cdf(kappa, Theta1) - vonmises.cdf(kappa, Theta0)` to calculate that measure.  Is that correct?

**Comment by josef-pkt** [comment]
> My interpretation of your earlier comments is that you may have used the difference vonmises.cdf(kappa, Theta1) - vonmises.cdf(kappa, Theta0) to calculate that measure. Is that correct?

Yes, because cdf of vonmises is trend + periodic term, only definite integrals make sense, i.e. there is no fixed origin with cdf = 0 that is relevant for any specific (shifted) problem. 
It's like going round and round the circle, but we only want to measure distances for the last circle, while our step counter (cdf) is measuring from some initial location.

**Comment by pvanmulbregt** [comment]
I wondered how tied `vonmises` is to [-pi, pi] so I created 2 other `vonmises_line` distributions with a longer interval, and a shifted interval:
`vonmises_line2  = vonmises_gen(a=-2*np.pi, b=2*np.pi, name='vonmises_line2')`
`vonmises_line3  = vonmises_gen(a=0, b=2*np.pi, name='vonmises_line3')`
analogous to the original
`vonmises_line = vonmises_gen(a=-np.pi, b=np.pi, name='vonmises_line')`
and plotted the CDFs.

![image](https://user-images.githubusercontent.com/23403152/40689646-1ba07cc2-6371-11e8-9f99-5b0891291dbb.png)

Looks like the only interval supported is `[-pi, pi]`. The length must be `2pi` and it must be centered at 0.  Any other interval would need to be handled by using `loc` and `scale`.

**Comment by pvanmulbregt** [comment]
My understanding from this is that a circular distribution framework would have:
- The PDF is repeating non-negative function with period `2pi`and `integral_{-pi}^{pi} PDF(x) = 1`
- The CDF is monotonically increasing, with `CDF(x+2pi) = CDF(x) +1`.  In particular, it will take negative values, and values bigger than 1. 
- The PPF's domain would not be restricted to `[0, 1]`. 
- Some "standard" functions in `rv_generic` wouldn't make a lot of sense in the circular framework.
Something like the plot below (vonMises with kappa=1)

![image](https://user-images.githubusercontent.com/23403152/40818010-15880630-6523-11e8-8fb3-8fb202e0d0be.png)
Is that correct?

**Comment by josef-pkt** [comment]
The first two are correct for sure. 
Aside to get one full rotation we would need only 2 cycles in the periodic pdf, cdf

I don't remember about the other two, I might not have really used rvs and ppf (I'm pretty sure I didn't look specifically at np.random.vonmises):

> The PPF's domain would not be restricted to [0, 1].
> Some "standard" functions in rv_generic wouldn't make a lot of sense in the circular framework.
Something like the plot below (vonMises with kappa=1)

My guess is to create rvs for vonmises we need a "proper" ppf, i.e. we should use vonmises_line and rotate, i.e. use a shift and a modulo transform, or transform ppf to map from [0, 1] to some interval of length 2 * pi if we want the generic `rvs = ppf(rvs_uniform)` transformation.

**Comment by pvanmulbregt** [comment]
I meant `rv_generic` in a different sense.  The class hierarchy is
```
rv_frozen
rv_generic ---- rv_discrete --- binom_gen
        |                |----- bernoulli_gen
        |                |----- ...
        |
        |------ rv_continuous --- norm_gen
                           |----- beta_gen
                           |----- ...

multi_rv_frozen
multi_rv_generic  --- multivariate_normal_gen
              | ----- dirichlet_gen
```
I was thinking about where to place `rv_circular` and what functionality it would have.
`rv_generic` contains useful functionality relating to arg checking and frozen distributions, but also includes methods which may not make sense for circular distributions.  E.g `_logcdf`, `_sf`, `_logsf`, `_isf`, `_stats`, `_stats`.  Others may have potentially different meanings `mean`, `std`, `var`. It may even need its own `rv_circular_frozen` as `rv_frozen` exposes `sf`, `stats` etc.
`rv_circular` could be a derived class of `rv_generic` with many methods set to `None`. It could be its own class with methods cloned from `rv_generic`.  Or pull the most generic functionality out of `rv_generic` into `rv_base` and have `rv_generic` derive from that. 

What circular distributions do you see needing in practice?
Uniform? von Mises? Wrapped Normal? Wrapped Cauchy? Wrapped Levy? All these?  Others?
Just 1-dimensional? Or on higher dimensional spheres (or even tori)?  [It looks like these higher dimensional cases are sufficiently different to treat separately and not consider at this time.]

**Comment by josef-pkt** [comment]
AFAIR, I never thought about what a full circular distribution class should look like, I was mainly working with pdf, cdf and characteristic function.

Higher dimensional spheres would require multivariate distributions and not fit directly into the same setting as univariate circular distributions.

AFAIR, the main intuition for implementing parts of the distribution is to map it to a fixed interval, e.g. [0, 2*pi] and translate everything into that space. 
We still need an origin to fix the meaning of the measurements, and observations are usually in the interval, E.g. a 24 hour clock starts at 0 and ends at 24  (excluding 24), time is measured on this scale, and the distribution is over those 24 hours.
The main point to take into account is that distances are not on the interval, e.g. 23.5 is close to 0.5, and we don't want a jump at 0. 
ppf could be defined on this fixed interval, e.g. for use in rvs.
However, the median is defined differently and would not be ppf(0.5) in this definition.
isf can be defined in the same way as cdf but with a negative trend, i.e. only definite integrals to a fixed endpoint make sense.

The other option is to define everything possible in terms of periodic functions, e.g. pdf and moments are defined in terms of complex exponential or sin/cos.

Aside: besides full circular data, data can also be undirected in the sense that we don't know if it's looking forward or backward, e.g. we cannot distinguish 3am from 3pm if we only know it's 3 o'clock in the US. In the application that we worked on, we had data for 0 to 180 degrees (which is equivalent to having 180 to 360 degree data, objects didn't have distinguishable head and tail).

summary: 
I was thinking in pretty eclectic ways about this, using for each case whatever was convenient or made sense to me, but I didn't have to bring everything into a consistent implementation. i.e. I was jumping between periodic functions and working with modulo on an interval pretty "arbitrarily".

**Comment by josef-pkt** [comment]
I think for designing circular distribution it would be helpful to look at astropy or ask those developers. Astronomy is one of the main application areas for circular distributions that I know of.
(but they also use `units` to handle the measurement scale and conversion, which wouldn't work for scipy)

**Comment by chrisb83** [comment]
I read a little bit about circular statistics and agree that a separate class would be needed in view of the different concept.

As a short-term fix, one could improve the docstring to point out that `vonmises` is not a true distribution etc

(I still don´t see much added value of `vonmises`: if I understand correctly, the only advantage is that one can use arbitrary intervals of length 2 * pi to evaluate the pdf/cdf. For the cdf, the user has to fix the left endpoint manually and in my view, this is not much more convenient than mapping data point to [-pi,pi]. so I would even consider removing it from the list of continuous random variables in scipy)

**Comment by chrisb83** [comment]
> The doc is also the same for both distributions, which is confusing to a reader. But perhaps not surprising given scipy/stats/_continuous_distns.py:6156-6157

Ref to #6358

**Comment by josef-pkt** [comment]
> this is not much more convenient than mapping data point to [-pi,pi]

If I understand what you mean, then no. Using a fixed interval requires that vonmises has two shape parameters that moves the underlying periodic pdf and cdf into the [-pi, pi] or [0, 2*pi] interval. Then we can shift datapoints into the same fixed interval.

**Comment by chrisb83** [comment]
@josef-pkt I intend to update a few of the docstrings of distribution in stats and wanted to add a few sentences on `vonmises` nd `vonmises_line` as well.

> `vonmises` is a circular distribution. Currently, there is no circular distribution framework in scipy. The `cdf` is implemented such that `cdf(x + 2*np.pi) == cdf(x) + 1`.
> 
> `vonmises_line` is the same distribution, defined on :math:`[-\pi, \pi]` on the real line. This is a regular (i.e. non-circular) distribution.

Any suggestions what else one should mention? Thanks for your input.

**Comment by josef-pkt** [comment]
(I don't remember the details right now for the different methods)
about vonmises: It should explain, if it doesn't already, that vonmises does not define and enforce bounds or a fixed interval for the support of the distribution. 
Users can choose an interval of length 2 pi for which vonmises can then be treated as a proper distribution (e.g. assuming we subtract cdf of lower bound.)
pdf(x) == pdf(x + 2 * pi) ?

(I don't remember how ppf is defined.)
